hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java

```diff
@@ -221,16 +225,63 @@ public TransactionContext startTransaction(
 
       // Add the term index and transaction log index to applyTransaction map
       // . This map will be used to update lastAppliedIndex.
+
+      CompletableFuture<Message> ratisFuture =
+          new CompletableFuture<>();
+
       applyTransactionMap.put(trxLogIndex, trx.getLogEntry().getTerm());
-      CompletableFuture<Message> future = CompletableFuture.supplyAsync(
-          () -> runCommand(request, trxLogIndex),
-          executorService);
-      return future;
-    } catch (IOException e) {
+      CompletableFuture<OMResponse> future = CompletableFuture.supplyAsync(
+          () -> runCommand(request, trxLogIndex), executorService);
+      future.thenApply(omResponse -> {
+        if(!omResponse.getSuccess()) {
+          // When INTERNAL_ERROR or METADATA_ERROR it is considered as
+          // critical error and terminate the OM. Considering INTERNAL_ERROR
+          // also for now because INTERNAL_ERROR is thrown for any error
+          // which is not type OMException.
+
+          // Not done future with completeExceptionally because if we do
+          // that OM will still continue applying transaction until next
+          // snapshot. So in OM case if a transaction failed with un
+          // recoverable error and if we wait till snapshot to terminate
+          // OM, then if some client requested the read transaction of the
+          // failed request, there is a chance we shall give wrong result.
+          // So, to avoid these kind of issue, we should terminate OM here.
+          if (omResponse.getStatus() == INTERNAL_ERROR) {
+            terminate(omResponse, OMException.ResultCodes.INTERNAL_ERROR);
+          } else if (omResponse.getStatus() == METADATA_ERROR) {
+            terminate(omResponse, OMException.ResultCodes.METADATA_ERROR);
+          } else {
+            // For all other errors which are not critical, we can complete
+            // future normally.
+            ratisFuture.complete(OMRatisHelper.convertResponseToMessage(
+                omResponse));
+          }
+        } else {
+          ratisFuture.complete(OMRatisHelper.convertResponseToMessage(
+              omResponse));
+        }
+        return ratisFuture;
+      });
+      return ratisFuture;
+    } catch (Exception e) {
       return completeExceptionally(e);
     }
   }
 
+  /**
+   * Terminate OM.
+   * @param omResponse
+   * @param resultCode
+   */
+  private void terminate(OMResponse omResponse,
+      OMException.ResultCodes resultCode) {
+    OMException exception = new OMException(omResponse.getMessage(),
+        resultCode);
+    String errorMessage = "OM Response has received unrecoverable " +
+        "error, terminating OM. Error Response is" + omResponse;
+    ExitUtils.terminate(1, errorMessage, exception, LOG);
+  }
```

Comment:
This is a critical error message. Can we expand on the error message to give as much information as possible? I am thinking we should mention that we received an unrecoverable error from OM Ratis server and to avoid DB corruption, we are terminating the OM. 
One question is do we have to terminate here in case of single node OM ratis also?

Reply:
> One question is do we have to terminate here in case of single node OM ratis also?
Good question, I am also not completely sure, but looks like in single node looks like we can continue. If it is really critical error like rocksdb error due to disk failure, anyway it will fail the user operations. But I am not completely sure which will be the correct way for single node OM. Thoughts?

Reply:
Addressed other comment on changing the log message.